### PR TITLE
OTLP for Metrics 

### LINKED ISSUE #1163: OTEL Metrics Collection : Batching & Label Processing
Related to [OEP](https://github.com/rancher/opni/blob/main/enhancements/monitoring/20230117-move-to-otel-collector.md)

### LINKED ISSUE #1193: OTEL Metrics Collection : OTEL GRPC & Gateway Remote Write

### PR DESCRIPTION
**Pull request by alexandreLamarre**
Closes #1193 

- [x] OTLP metrics between gateway and agent

### CODE REVIEW

File: plugins/metrics/pkg/backend/metrics.go [hunk 1]
```diff
@@ -107,6 +109,45 @@ func (m *MetricsBackend) requestNodeSync(ctx context.Context, cluster *corev1.Re
...
+		}
+		for _, cap := range cl.Metadata.Capabilities {
+			if cap.Name == wellknown.CapabilityMetrics {
+				spec, err := m.GetNodeConfiguration(ctx, cl.Reference())
+				if err != nil {
+					m.Logger.Warnf("failed to get node configuration for cluster %s: %v", cl.Reference().GetId(), err)
+					continue
+				}
+				if spec != nil { // TODO : #1176 check for otel capability
+					shouldDeployCollector = true
```
Comment (dbason) [inline plugins/metrics/pkg/backend/metrics.go:143]: Do we actually do anything with this value?

**Review by dbason — COMMENTED** [review]
Just one question about the syncing

File: plugins/metrics/pkg/backend/metrics.go (hunk 1 above)
Comment (alexandreLamarre) [reply plugins/metrics/pkg/backend/metrics.go:143]: it's captured in the defer func in order to tell the cluster driver how to deploy otel gateway resources
https://github.com/rancher/opni/blob/c8469d078f588c0cbc4ee769ce1bd756b6e5b87e/plugins/metrics/pkg/backend/metrics.go#L114-L120

Comment (alexandreLamarre) [reply plugins/metrics/pkg/backend/metrics.go:143]: I think this function is hard to read though, so let me rework this a little

File: plugins/metrics/pkg/agent/stream.go [hunk 2]
```diff
@@ -21,18 +22,23 @@ func (p *Plugin) StreamServers() []streamext.Server {
 			Desc: &remoteread.RemoteReadAgent_ServiceDesc,
 			Impl: p.node,
 		},
+		{
```
[the remark this answers is not in the document]
Comment (alexandreLamarre) [reply plugins/metrics/pkg/agent/stream.go:25]: Requires https://github.com/dbason/featureflags/pull/1

[the remark this answers is not in the document]
Comment (alexandreLamarre) [reply plugins/metrics/pkg/agent/stream.go:25]: Will be "feature gated" by a node.proto `useOTLP` flag

**Review by kralicky — COMMENTED** [review]
This PR is **on hold** and should be re-opened closer to 1.0 after relevant major features and refactors are complete.

File: pkg/test/testdata/testdata/otel/otlp-agent-forwarder.tmpl [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+          static_configs:
+            - targets: [':{{ .TelemetryPort }}']
+processors:
+  resourcedetection/system:
+    detectors: ["system"]
+    system:
+      hostname_sources: ["lookup"]
+exporters:
+  {{ if .OTLPExporterEndpoint }}
+  otlp:
```
Comment (jaehnri) [inline pkg/test/testdata/testdata/otel/otlp-agent-forwarder.tmpl:16]: Locally, I couldn't get `forwarder_test.go` to work: 
```
  2023-09-29T22:29:14.078Z	error	exporterhelper/queued_retry.go:401	Exporting failed. The error is not retryable. Dropping data.	{"kind": "exporter", "data_type": "metrics", "name": "otlp", "error": "Permanent error: rpc error: code = Unimplemented desc = grpc: Decompressor is not installed for grpc-encoding \"gzip\"", "dropped_items": 16}
```

The fix was to append `compression: none` to the OTLP exporter.
```suggestion
  otlp:
    compression: none    
```

**Comment by alexandreLamarre** [comment]
Out of date

**Pull request closed without merging**